x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/DataNodeRequestSender.java

```diff
@@ -0,0 +1,336 @@
+        } finally {
+            listener.onResponse(null);
+        }
+    }
+
+    private void sendOneNodeRequest(TargetShards targetShards, ComputeListener computeListener, NodeRequest request) {
+        final ActionListener<List<DriverProfile>> listener = computeListener.acquireCompute();
+        sendRequest(request.node, request.shardIds, request.aliasFilters, new NodeListener() {
+            void onAfter(List<DriverProfile> profiles) {
+                nodePermits.get(request.node).release();
+                trySendingRequestsForPendingShards(targetShards, computeListener);
+                listener.onResponse(profiles);
+            }
+
+            @Override
+            public void onResponse(DataNodeComputeResponse response) {
+                // remove failures of successful shards
+                for (ShardId shardId : targetShards.shardIds()) {
+                    if (shardFailures.containsKey(shardId) == false) {
+                        shardFailures.remove(shardId);
+                    }
+                }
+                for (Map.Entry<ShardId, Exception> e : response.shardLevelFailures().entrySet()) {
+                    final ShardId shardId = e.getKey();
+                    shardFailures.compute(shardId, (k, v) -> mergeShardExceptions(v, e.getValue()));
+                    pendingShardIds.add(shardId);
+                }
+                onAfter(response.profiles());
+            }
+
+            @Override
+            public void onFailure(Exception e, boolean receivedData) {
+                if (receivedData) {
+                    aborted.set(true);
+                }
+                for (ShardId shardId : request.shardIds) {
+                    shardFailures.compute(shardId, (k, v) -> mergeShardExceptions(v, e));
+                    pendingShardIds.add(shardId);
+                }
+                onAfter(List.of());
+            }
+        });
+    }
+
+    abstract void sendRequest(DiscoveryNode node, List<ShardId> shardIds, Map<Index, AliasFilter> aliasFilters, NodeListener nodeListener);
+
+    interface NodeListener {
+        void onResponse(DataNodeComputeResponse response);
+
+        void onFailure(Exception e, boolean receivedData);
+    }
+
+    private static Exception mergeShardExceptions(Exception current, Exception e) {
+        e = e instanceof TransportException te ? FailureCollector.unwrapTransportException(te) : e;
+        if (TransportActions.isShardNotAvailableException(e)) {
+            e = NoShardAvailableActionException.forOnShardFailureWrapper(e.getMessage());
+        }
+        if (current == null) {
+            return e;
+        }
+        if (current instanceof NoShardAvailableActionException || ExceptionsHelper.unwrap(current, TaskCancelledException.class) != null) {
+            return e;
+        }
+        if (e instanceof NoShardAvailableActionException || ExceptionsHelper.unwrap(e, TaskCancelledException.class) != null) {
+            return current;
+        }
```

Comment:
Should we keep the other exception as a suppressed one? It might be nice to know that we got canceled and then retried.

Reply:
I added a comment to explain why we retain only one meaningful exception and avoid suppressing previous failures.